### PR TITLE
Cultural Invariant Conversion of datatypes during mutation

### PR DESCRIPTION
**Pull request by sourabh1007**
## Why make this change?

It addresses https://github.com/Azure/data-api-builder/issues/2284 bug raised by the customer.

**Note:** Made change in`src/Core/Resolvers/SqlMutationEngine.cs` just for consistency.

## What is this change?
Made sure datatype conversion is `CultureInfo.InvariantCulture`

## How was this tested?

- [ ] Integration Tests
- [ ] Unit Tests

## Sample Request(s)

- Example REST and/or GraphQL request to demonstrate modifications
- Example of CLI usage to demonstrate modifications

### CODE REVIEW

**Comment by seantleonard** [comment]
please explain why this works, what's the behavior before/after.

**Comment by sourabh1007** [comment]
> please explain why this works, what's the behavior before/after.

you just want me to put in description or it is a question, because changes are not clear enough? I have test cases added for this, you can refer that, that case will fail if you run it with main branch.
To know the "before" behavior, you can check the mentioned issue.

File: src/Core/Resolvers/SqlMutationEngine.cs
Comment (seantleonard) [inline src/Core/Resolvers/SqlMutationEngine.cs:1]: why the changes in this file? The linked issue and the tests in this pr only test Cosmos. Not sure why sql changes are made and why the changes were made to the "records affected" integer.. that value isn't returned to users.

File: src/Core/Services/TypeHelper.cs [hunk 1]
```diff
@@ -304,12 +305,12 @@ public static bool TryGetDbTypeFromSqlDbDateTimeType(SqlDbType sqlDbType, [NotNu
             SyntaxKind valueKind = node.Kind;
             return valueKind switch
             {
-                SyntaxKind.IntValue => Convert.ToInt32(node.Value), // spec
-                SyntaxKind.FloatValue => Convert.ToDouble(node.Value), // spec
-                SyntaxKind.BooleanValue => Convert.ToBoolean(node.Value), // spec
-                SyntaxKind.StringValue => Convert.ToString(node.Value), // spec
+                SyntaxKind.IntValue => Convert.ToInt32(node.Value, CultureInfo.InvariantCulture), // spec
+                SyntaxKind.FloatValue => Convert.ToDouble(node.Value, CultureInfo.InvariantCulture), // spec
+                SyntaxKind.BooleanValue => Convert.ToBoolean(node.Value, CultureInfo.InvariantCulture), // spec
```
Comment (seantleonard) [inline src/Core/Services/TypeHelper.cs:310]: why do you need to define `CultureInfo.InvariantCulture` for Convert.ToBoolean? that parameter is ignored per documentation:https://learn.microsoft.com/dotnet/api/system.convert.toboolean?view=net-8.0#system-convert-toboolean(system-string-system-iformatprovider)

File: src/Core/Services/TypeHelper.cs [hunk 2]
```diff
@@ -304,12 +305,12 @@ public static bool TryGetDbTypeFromSqlDbDateTimeType(SqlDbType sqlDbType, [NotNu
             SyntaxKind valueKind = node.Kind;
             return valueKind switch
             {
-                SyntaxKind.IntValue => Convert.ToInt32(node.Value), // spec
-                SyntaxKind.FloatValue => Convert.ToDouble(node.Value), // spec
-                SyntaxKind.BooleanValue => Convert.ToBoolean(node.Value), // spec
-                SyntaxKind.StringValue => Convert.ToString(node.Value), // spec
+                SyntaxKind.IntValue => Convert.ToInt32(node.Value, CultureInfo.InvariantCulture), // spec
+                SyntaxKind.FloatValue => Convert.ToDouble(node.Value, CultureInfo.InvariantCulture), // spec
```
Comment (seantleonard) [inline src/Core/Services/TypeHelper.cs:309]: Would be helpful to know **why** this is the correct change to make (why here?).
- Is this a breaking change? 
- does this change how values are stored in the backend database? The tests don't demonstrate this.
- examples from a chart like this would help. 

|Provided Value | client culture| culture | persisted value | returned value|
|- |- |- |- |- |
|12.32 (double)| en-DE | InvariantCulture| 12,32 | 12,32 |
|12.32 (double)| en-DE | Before change, no culture specified| 12.32| 1232??|

File: src/Service.Tests/CosmosTests/MutationTests.cs [hunk 3]
```diff
@@ -1074,6 +1075,31 @@ public async Task CanPatchMoreThan10AttributesInAnItemWithVariables()
             Assert.AreEqual(patchResponse.GetProperty("moons").ToString(), JsonSerializer.Serialize(update.moons));
         }
 
+        [TestMethod]
+        [DataRow("en-US")]
+        [DataRow("en-DE")]
+        public async Task CanCreateItemWithCultureInvariant(string cultureInfo)
+        {
+            CultureInfo ci = new(cultureInfo);
+            CultureInfo.DefaultThreadCurrentCulture = ci;
```
Comment (seantleonard) [inline src/Service.Tests/CosmosTests/MutationTests.cs:1084]: Comment would help, this defines the culture of the DAB "host" machine or "thread." 

So in this test a double value `10.15` (en-US culture) would be persisted to database as "cultureInfo.InvariantCulture" which wouldn't change the value -> stored as `10.15`. but then when response is returned, because the DefaultThreadCulture is `en-DE`, the value is converted in the response to match the host or client's culture?

**Review by seantleonard — CHANGES_REQUESTED** [review]

File: src/Core/Resolvers/SqlMutationEngine.cs
Comment (sourabh1007) [reply src/Core/Resolvers/SqlMutationEngine.cs:1]: I am reverting this change. 
By the way, it doesn't matter if you are returning this value to the customer or not. This code and cosmos DB code which I am fixing, both runs on the server and so issue can happen here also, So I thought to put it as safety check. 
But you are right, it needs to be tested separately. SQL team can take it up if there is any issue in the future due to this.

File: src/Core/Services/TypeHelper.cs (hunk 2 above)
Comment (sourabh1007) [reply src/Core/Services/TypeHelper.cs:309]: In the attached issue, it is clearly mentioned that, when customer is trying to save 12.32, it is getting converted to 1232 and then same is returning back.
1) I don't think it is a breaking change. It is a fix. ideally whatever you are sending that should be saved.
2) Yes, in the issue it is already mentioned
image
3) it not about `en-DE` only, it about general application behavior, application should not take the decision of converting values. it should save whatever is coming from customer.

File: src/Service.Tests/CosmosTests/MutationTests.cs (hunk 3 above)
Comment (sourabh1007) [reply src/Service.Tests/CosmosTests/MutationTests.cs:1084]: No, In response, value won't be converted back to host/client culture, thats what I am asserting here by putting "expected value" as cultural invariant.

**Review by neeraj-sharma2592 — APPROVED** [review]

**Review by seantleonard — APPROVED** [review]
